### DISCUSSION TITLE
Workspace-specific taskbar

### DISCUSSION THREAD

**Discussion by ATM-Jahid** [hyprwm/Hyprland#1094] [General]
Does anybody use a taskbar that groups tasks according to workspaces?

I was trying to achieve this using waybar. However, the `wlr/taskbar` module does not have that ability [(Waybar#1278)](https://github.com/Alexays/Waybar/issues/1278).

Would it be possible to create a module like that for Hyprland (something like `hyprland/ws-taskbar`)?

**Comment by vaxerski** [comment]
You'd have to script your own I think.

**Comment by Person1873** [comment]
I'm working on a script for this, only in the early stages, but should have something soon (TM)
I'll share what i have so far though


https://github.com/Person1873/hypr-taskbar

```
#!/bin/bash
#script depends on below packages
#hyperland (duh)
#waybar (double duh)
#jq
#ripgrep (can be substituted for normal grep)

WORKSPACE=$(
	hyprctl activewindow |
	rg workspace |
	awk '{print $2}');

MONITOR=$(
	hyprctl activewindow |
	rg monitor |
	awk '{print $2}');

echo workspace: $WORKSPACE;
echo monitor: $MONITOR;

hyprctl clients -j |
jq -r --arg WORKSPACE "$WORKSPACE" 'map(select(.workspace.id == ($WORKSPACE|tonumber)))';
```

**Reply by Person1873** [reply]
I've hit a dead-end on this approach.
I will actually need to patch waybar directly, however there is some promise here for simply showing the currently active window.

**Comment by yavko** [comment]
Ironbar has support for this with Hyprland out of box https://github.com/JakeStanger/ironbar

**Reply by ATM-Jahid** [reply]
Nice bar! I see that there are two related modules: Launches and Workspaces. However, I don't see the option to group programs according to workspaces. Am I missing something here?

**Reply by yavko** [reply]
> Nice bar! I see that there are two related modules: Launches and Workspaces. However, I don't see the option to group programs according to workspaces. Am I missing something here?

I'm not entirely sure if this is supported actually, I misread the issue, sorry.

**Comment by Person1873** [comment]
It would appear that the best solution for now is to either not use waybar for listing windows (only for a status bar) and use rofi/wofi/dmenu/fzf etc as a switcher.
there are a few compatible scripts available here
https://www.reddit.com/r/hyprland/comments/113ofeq/script_with_rofi_to_see_open_apps_and_switch_to/
it would be possible to filter this by workspace using the script i posted earlier but with some modifications

**Comment by Azarattum** [comment]
Also looking for this. Would be really nice to have workspaces and tasks as a single widget (where windows are grouped by workspace and clicking the icon brings you to that workspace and focuses the window). Do any workarounds exist currently? It's been a year since last comment here...

**Reply by ATM-Jahid** [reply]
I tried to make [my bar with Qt](https://github.com/ATM-Jahid/qtventure). However, I gave up because there are decade-old bugs in Qt causing flickers during window resize events.

This was in December when I had some free time. If I try to make one again, I'll probably go with Flutter since they implemented the [layer shell protocol](https://pub.dev/packages/wayland_layer_shell).

FlafyDev has a [working bar here](https://github.com/flafydev/flutter_background_bar). You can see the demo in [this Reddit post](https://www.reddit.com/r/unixporn/comments/13gz5kq/hyprland_my_summer_rice/).

**Reply by yavko** [reply]
> I tried to make [my bar with Qt](https://github.com/ATM-Jahid/qtventure). However, I gave up because there are decade-old bugs in Qt causing flickers during window resize events.
> 
> This was in December when I had some free time. If I try to make one again, I'll probably go with Flutter since they implemented the [layer shell protocol](https://pub.dev/packages/wayland_layer_shell).
> 
> FlafyDev has a [working bar here](https://github.com/flafydev/flutter_background_bar). You can see the demo in [this Reddit post](https://www.reddit.com/r/unixporn/comments/13gz5kq/hyprland_my_summer_rice/).

Try quickshell, it allows you to build a shell using qml, and works really well

**Reply by ATM-Jahid** [reply]
> Try quickshell, it allows you to build a shell using qml, and works really well

The flickering problem during resize is inherent to QML. How is quickshell gonna solve that?

**Reply by yavko** [reply]
> > Try quickshell, it allows you to build a shell using qml, and works really well
> 
> The flickering problem during resize is inherent to QML. How is quickshell gonna solve that?
> 

Idk I've never experienced it

**Comment by NikkSaan** [comment]
[nwg-panel](https://github.com/nwg-piotr/nwg-panel) supports grouping windows by workspace.

**Comment by fdecarpentier** [comment]
I think it's possible out of the box with waybar. See here https://www.lorenzobettini.it/2024/08/waybar-workspaces-in-hyprland/#google_vignette

**Reply by ATM-Jahid** [reply]
I do use waybar in that way. Here's [my config](https://github.com/ATM-Jahid/afrodots/tree/main/fakehome/.config/waybar). What doesn't work is focusing on a particular window or moving it to a different workspace by clicking. I'm fine with my keyboard shortcuts though. I just wanted to recreate the functionality of my old bspwm+tint2 setup.

**Comment by ArielBaron** [comment]
probably already said but there is built in taskbar module for waybar: for example:
 "hyprland/workspaces": {
      "format": "{icon}: {windows}",
      "format-window-separator": " | ",
      "workspace-taskbar": {
        "enable": true,
        "update-active-window": true,
        "format": "{icon} {title:.22}",
        "icon-size": 18,
        "on-click-window": "${SCRIPTS}/focus-window.sh {address} {button}"
      }
    },

**Comment by Finalet** [comment]
If you want to display app icons, it was impossible out of the box, even with regular "custom/..." modules. So, I vibe coded my own module in C using the [CFFI module](https://github.com/Alexays/Waybar/wiki/Module:-CFFI).

Check it out here: [hypr-ws-apps](https://github.com/Finalet/Linux-Configs/tree/main/waybar/hypr-ws-apps)

In short, you can create a config and specify from which workspace to display apps. Either provide a workspace index (number) or name (string).

```json
{
  "modules-center": ["cffi/hypr-ws-apps"],
  "cffi/hypr-ws-apps": {
    "module_path": "/home/<user>/.config/waybar/cffi/libhypr_ws_apps.so",
    "workspace": "1",
    "icon_size": 16,
    "spacing": 8,
    "max_icons": 12,
    "show_empty": false,
    "tooltip": true
  }
}
```

You can download my `.so` file and plug it into your CFFI module to play around.
